modules/guides/src/GuidesStorage.php

```diff
@@ -0,0 +1,158 @@
+
+  /**
+   * The settings of the site.
+   *
+   * @var \Drupal\Core\Site\Settings
+   */
+  protected $settings;
+
+  /**
+   * The cache backend service.
+   *
+   * @var \Drupal\Core\Cache\CacheBackendInterface
+   */
+  protected $cacheBackend;
+
+  /**
+   * GuidesStorage constructor.
+   *
+   * @param \Drupal\Core\Site\Settings $settings
+   *   The settings of the site.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache_backend
+   *   The cache backend.
+   */
+  public function __construct(Settings $settings, CacheBackendInterface $cache_backend) {
+    $this->settings = $settings;
+    $this->cacheBackend = $cache_backend;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  private function getDirectory(): string {
+    return DRUPAL_ROOT . ($this->settings->get('guides_dir') ?? '/guides');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFilePaths(): array {
+    $directory = $this->getDirectory();
+    $cid = 'guides_files:' . $directory;
+    $data_cached = $this->cacheBackend->get($cid);
+
+    if ($data_cached) {
+      $guides = $data_cached->data;
+    }
+    else {
+      $iterator = new \RecursiveDirectoryIterator($directory, \FilesystemIterator::CURRENT_AS_FILEINFO | \FilesystemIterator::SKIP_DOTS);
+      $filter = new \RecursiveCallbackFilterIterator($iterator, static function (\SplFileInfo $current, string $key, \RecursiveIterator $iterator): bool {
+        if ($iterator->hasChildren()) {
+          return TRUE;
+        }
+        $path = $current->getPathname();
+        if (is_file($path)) {
+          return $current->getExtension() === 'md';
+        }
+
+        return FALSE;
+      });
+
+      $guides = [];
+      $files = new \RecursiveIteratorIterator($filter);
+      foreach ($files as $md) {
+        if ($files->getDepth() === 1) {
+          $guides[] = $md->getPathname();
+        }
+      }
+
+      $this->cacheBackend->set($cid, $guides);
+    }
+
+    return $guides;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFilePath(string $subdirectory, string $file_name): string {
+    $file_path = $this->getDirectory() . DIRECTORY_SEPARATOR . $subdirectory . DIRECTORY_SEPARATOR . $file_name . '.md';
+
+    if (in_array($file_path, $this->getFilePaths()) && file_exists($file_path)) {
+      return $file_path;
+    }
+    else {
+      // If the file does not exist, then the cache contains invalid data, so it
+      // has to be rebuilt.
+      $this->cacheBackend->deleteAll();
+      throw new FileNotFoundException($file_path);
+    }
+
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFileContent(string $subdirectory, string $file_name): string {
+    $file_path = $this->getFilePath($subdirectory, $file_name);
+    return str_replace('@guide_path', dirname($file_path), file_get_contents($file_path));
+
+  }
+
+  /**
+   * Gets the links of the guides files.
+   *
+   * This method intentionally does not validate whether a file still exists or
```

Comment:
Use `{@inheritdoc}` here and move this comment to method's documentation in the parent interface.